iree/compiler/InputConversion/MHLO/MHLOToLinalgOnTensors.cpp

```diff
@@ -207,7 +203,81 @@ struct FftOpConversion : public OpConversionPattern<mhlo::FftOp> {
     return success();
   }
 };
-}  // namespace
+
+// We need to convert func ops in order to convert types.
+class BuiltinFuncOpPattern : public OpConversionPattern<FuncOp> {
+  using OpConversionPattern<FuncOp>::OpConversionPattern;
+  LogicalResult matchAndRewrite(
+      FuncOp srcOp, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    FunctionType srcFuncType = srcOp.getType();
+    TypeConverter::SignatureConversion signatureConversion(
+        srcOp.getNumArguments());
+
+    // Convert function arguments.
+    for (unsigned i = 0, e = srcFuncType.getNumInputs(); i < e; ++i) {
+      if (failed(getTypeConverter()->convertSignatureArg(
+              i, srcFuncType.getInput(i), signatureConversion))) {
+        return rewriter.notifyMatchFailure(srcOp, "argument failed to convert");
+      }
+    }
+
+    // Convert function results.
+    SmallVector<Type, 1> convertedResultTypes;
+    if (failed(getTypeConverter()->convertTypes(srcFuncType.getResults(),
+                                                convertedResultTypes))) {
+      return rewriter.notifyMatchFailure(srcOp, "results failed to convert");
+    }
+
+    // Create new function with converted argument and result types.
+    // Note that attributes are dropped. Consider preserving some if needed.
+    auto newFuncType = mlir::FunctionType::get(
+        srcOp.getContext(), signatureConversion.getConvertedTypes(),
+        convertedResultTypes);
+
+    // Update the function in place.
+    rewriter.startRootUpdate(srcOp);
+    srcOp.setType(newFuncType);
+
+    // Tell the rewriter to convert the region signature.
+    TypeConverter &typeConverter = *getTypeConverter();
+    if (failed(rewriter.convertRegionTypes(&srcOp.getBody(), typeConverter,
+                                           &signatureConversion))) {
+      return failure();
+    }
+
+    rewriter.finalizeRootUpdate(srcOp);
+    return success();
+  }
+};
+
+class GenericTypeConvert : public ConversionPattern {
+ public:
+  GenericTypeConvert(StringRef rootName, TypeConverter &converter,
+                     MLIRContext *context, PatternBenefit benefit = 0)
+      : ConversionPattern(converter, rootName, benefit, context) {}
+  LogicalResult matchAndRewrite(
+      Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const override {
+    llvm::SmallVector<NamedAttribute, 4> newAttr;
+    llvm::append_range(newAttr, op->getAttrs());
+    llvm::SmallVector<Type, 4> newResults;
+    (void)getTypeConverter()->convertTypes(op->getResultTypes(), newResults);
```

Comment:
Do we really want to ignore failure here? Most type converters that I've seen (including the MHLO one) do a passthrough for all other types, so I expect a failure here would be an error (https://github.com/tensorflow/tensorflow/blob/38d26c0717f68255944c8183a586ad8e2af3effa/tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/type_conversion.cc#L73)

Reply:
copy-pasta. expanded to handle error.